src/net-to-gross/calcTaxableIncome.js

```diff
@@ -0,0 +1,98 @@
+const Big = require('big.js');
+const TAX_RATES = require('../taxRate');
+const salaryConstants = require('../salaryConstants');
+
+/**
+ * Taxable Income (ti) = net + totalTax - 11
+ * @type {[{taxableIncomeFactor: number, maxNet: number}]}
+ */
+const TAX_FACTOR_RANGES = [
+    // ti <= net + 5% *min(ti, 5) - 11
+    // ti <= (net - 11)/ 0.95
+    // ti <= 5 ==> net <= 15.75
+    {netMax: 15_750_000, taxableIncomeFactor: 0.95},
+
+    // ti <= net + 5% * 5 + 10% *min(ti - 5, 5) - 11
+    // ti <= (net - 11.25)/0.9
+    // ti <= 10 ==> net <= 20.25
+    {netMax: 20_250_000, taxableIncomeFactor: 0.9},
+
+    // ti <= net + 5% * 5 + 10% * 5 + 15%*min(ti - 10, 8) - 11
+    // ti <= (net - 11.75)/0.85
+    // ti <= 18 ==> net <= 27.05
+    {netMax: 27_050_000, taxableIncomeFactor: 0.85},
+
+    // ti <= net + 5% * 5 + 10% * 5 + 15%*8 + 20%*min(ti-18, 14)- 11
+    // ti <= (net -12.65)/0.8
+    // ti <= 32 ==> net <= 38.25
+    {netMax: 38_250_000, taxableIncomeFactor: 0.8},
+
+    // ti <= net + 5% * 5 + 10% * 5 + 15%*8 + 20%*14 + 25%*min(ti-32, 20) - 11
+    // ti <= (net -14.25)/0.75 Only if ti <= 52
+    // ti <= 52 ==> net <= 53.25
+    {netMax: 53_250_000, taxableIncomeFactor: 0.75},
+
+    // ti <= net + 5% * 5 + 10% * 5 + 15%*8 + 20%*14 + 25%*20 +30%*min(ti-52, 28) - 11
+    // ti <= (net -16.85)/0.7
+    // ti <= 80 ==> net <= 72.85
+    {netMax: 72_850_000, taxableIncomeFactor: 0.7},
+
+    // ti >= net + 5% * 5 + 10% * 5 + 15%*8 + 20%*14 + 25%*20 +30%*28 + 35%*min(ti-80) - 11
+    // ti >= (net -20.85)/0.65
+    // ti >= 80 ==> net >= 72.85
+    {netMax: Number.MAX_SAFE_INTEGER, taxableIncomeFactor: 0.65}
+]
+
+/**
+ * @param netIncome
+ * @param dependentCount
+ * @returns {Promise<unknown>}
+ */
+
+module.exports = function calcTaxableIncome(netIncome, dependentCount = 0) {
+    return new Promise(function (resolve, reject) {
+        const dependents = dependentCount * salaryConstants.DEDUCTION_PER_PERSON;
+
+        let netBig = new Big(netIncome);
+        let ti = new Big(0)
+        let totalDeductionTimesRatePrev = new Big(0);
+        let totalDeductionPrev = new Big(0);
+
+        if (netBig.lte(salaryConstants.SELF_DEDUCTION)) {
+            ti = 0;
+            resolve(ti);
+        }
+
+        const taxFactorRanges = TAX_FACTOR_RANGES.find(function (result) {
+            return netBig.lte(result.netMax);
+        });
+        const taxableIncomeFactor = taxFactorRanges.taxableIncomeFactor
+
+        let taxRateMatch;
+        for (let i = 0; i < TAX_RATES.length; i++) {
+            const taxRate = TAX_RATES[i];
+            if ((1 - taxRate.rate) >= taxableIncomeFactor) {
```

Comment:
` taxRate.rate` is a decimal number. Will the subtraction `(1 - taxRate.rate)` makes any number error?